src/CanvasInput/index.js

```diff
@@ -109,58 +96,46 @@ export default class CanvasInput extends React.Component {
       hoverData: solvedData.hoverData,
       layerProps: solvedData.layerProps,
     })
-    this.lastMouse = mouse
   }
 
-  handleMouseMove = evt => {
-    const mouse = getMouseFromEvent(evt)
+  handleMouseMove = event => {
+    const mouse = getMouseFromEvent(event)
     const solvedData = runHoverSolverOn(
       getDataUnderMouse(this.props, mouse, this.canvasRef.current)
     )
-    let mouseDelta
-    if (this.lastMouse) {
-      mouseDelta = {
-        x: this.lastMouse.x - mouse.x,
-        y: this.lastMouse.y - mouse.y,
+    this.setState(prevState => {
+      const mouseDelta = prevState.mouse
+        ? {x: prevState.mouse.x - mouse.x, y: prevState.mouse.y - mouse.y}
+        : null
+      this.props.onUpdate({
+        action: prevState.mouseDrag ? 'mouseDrag' : 'mouseMove',
+        mouse,
+        mouseDelta,
+        ...solvedData,
+        rootProps: this.props.rootProps,
+      })
+      return {
+        mouseDrag: !!prevState.mouseDown,
+        mouse,
+        hoverRenderData: solvedData.hoverRenderData,
+        hoverData: solvedData.hoverData,
+        layerProps: solvedData.layerProps,
+        mouseLeave: false,
       }
```

Comment:
Same issue here, though it's not as straightforward to solve. Maybe look at using the second arg of `setState`: a callback function to execute after the state update.

Reply:
Done!